runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,406 @@
+    let mut ring = new_ring(&cfg).expect("unable to create io_uring instance");
+    let mut next_work_id: u64 = 0;
+    // Maps a work ID to the sender that we will send the result to.
+    let mut waiters: std::collections::HashMap<_, oneshot::Sender<i32>> =
+        std::collections::HashMap::with_capacity(cfg.size as usize);
+
+    loop {
+        // Try to get a completion
+        if let Some(cqe) = ring.completion().next() {
+            let work_id = cqe.user_data();
+            let result = cqe.result();
+            let sender = waiters.remove(&work_id).expect("work is missing");
+            // Notify with the result of this operation
+            let _ = sender.send(result);
+            continue;
+        }
+
+        // Try to fill the submission queue with incoming work.
+        // Stop if we are at the max number of processing work.
+        while waiters.len() < cfg.size as usize {
+            // Wait for more work
+            let (mut work, sender) = if waiters.is_empty() {
+                // Block until there is something to do
+                match receiver.next().await {
+                    Some(work) => work,
+                    None => return,
+                }
+            } else {
+                // Handle incoming work
+                match receiver.try_next() {
+                    // Got work without blocking
+                    Ok(Some(work_item)) => work_item,
+                    // Channel closed, shut down
+                    Ok(None) => return,
+                    // No new work available, wait for a completion
+                    Err(_) => break,
+                }
+            };
+
+            // Assign a unique id
+            let work_id = next_work_id;
+            work = work.user_data(work_id);
+            // Use wrapping add in case we overflow
+            next_work_id = next_work_id.wrapping_add(1);
+
+            // We'll send the result of this operation to `sender`.
+            waiters.insert(work_id, sender);
+
+            // Submit the operation to the ring
+            unsafe {
+                ring.submission()
+                    .push(&work)
+                    .expect("unable to push to queue");
+            }
+        }
+
+        // Wait for at least 1 item to be in the completion queue.
+        // Note that we block until anything is in the completion queue,
+        // even if it's there before this call. That is, a completion
+        // that arrived before this call will be counted and cause this
+        // call to return. Note that waiters.len() > 0 here.
+        ring.submit_and_wait(1).expect("unable to submit to ring");
+    }
+}
+
+#[derive(Clone)]
+pub struct Storage {
+    storage_directory: PathBuf,
+    io_sender: mpsc::Sender<(SqueueEntry, oneshot::Sender<i32>)>,
+}
+
+impl Storage {
+    /// Returns a new `Storage` instance.
+    /// The `Spawner` is used to spawn the background task that handles IO operations.
+    pub fn start(cfg: &Config) -> Self {
```

Comment:
nit: why not take in `Config`? Feels like we shouldn't need all these inner clones.